jobserver/serializers.py

```diff
@@ -22,57 +23,79 @@ class Meta:
+    """
+
     url = serializers.HyperlinkedIdentityField(view_name="jobs-detail")
+    pk = serializers.IntegerField(read_only=True)
+    backend = serializers.CharField(source="request.backend")
+    started = serializers.BooleanField(default=False)
+    force_run = serializers.BooleanField(default=False)
+    force_run_dependencies = serializers.BooleanField(
+        source="request.force_run_dependencies", default=False
+    )
+    action_id = serializers.CharField()
+    status_code = serializers.IntegerField(allow_null=True, required=False)
+    status_message = serializers.CharField(allow_null=True, required=False)
     outputs = JobOutputSerializer(many=True, required=False)
-    workspace = WorkspaceSerializer(read_only=True)
-    workspace_id = serializers.PrimaryKeyRelatedField(
-        queryset=Workspace.objects.all(), source="workspace"
+    needed_by_id = serializers.IntegerField(allow_null=True)
+    workspace = WorkspaceSerializer(read_only=True, source="request.workspace")
+    workspace_id = serializers.IntegerField(
+        source="request.workspace_id", required=False
     )
-    needed_by_id = serializers.PrimaryKeyRelatedField(
-        queryset=Job.objects.all(), source="needed_by", required=False
+    created_at = serializers.DateTimeField(read_only=True)
+    started_at = serializers.DateTimeField(read_only=True)
+    completed_at = serializers.DateTimeField(read_only=True)
+    callback_url = serializers.CharField(
+        source="request.callback_url", allow_null=True, required=False
     )
 
-    class Meta:
-        model = Job
-        fields = [
-            "url",
-            "pk",
-            "backend",
-            "started",
-            "force_run",
-            "force_run_dependencies",
-            "action_id",
-            "status_code",
-            "status_message",
-            "outputs",
-            "needed_by_id",
-            "workspace",
-            "workspace_id",
-            "created_at",
-            "started_at",
-            "completed_at",
-            "callback_url",
-        ]
-        read_only_fields = ["pk", "url", "created_at", "started_at", "completed_at"]
-
     def create(self, validated_data):
-        outputs_data = validated_data.pop("outputs", [])
-        job = Job.objects.create(**validated_data)
-        for output_data in outputs_data:
-            JobOutput.objects.create(job=job, location=output_data)
+        # walk the dep links until we reach the leaf Job, which we expect to
+        # have been created by job-server.  We can get a JobRequest from this
+        # to attach new Jobs to.
+        try:
+            dependency = Job.objects.get(pk=validated_data["needed_by_id"])
+        except Job.DoesNotExist:
+            raise ValidationError("needed_by_id must be a valid Job ID")
+
+        while dependency.needed_by is not None:
+            dependency = dependency.needed_by
+
+        job_request = dependency.request
```

Comment:
Reading this, I got confused, and it makes me realise `request` is perhaps a bad name for this relation (`request` being a particularly overloaded term in Django...)

I wonder if the relation should actually be `job_request`, for clarity... (sorry, should have thought of this in an earlier PR!)

Reply:
Arg, I know, I keep flip flopping on this.  I really dislike overly specific model or relation names (job.job_foo.job_bar, etc), but you're right about `request` here, just reading the snippet of code I wrote yesterday is already confusing, a good sign it needs to change!